File: cpp/1032.cpp
#include <iostream>
#include <cstring>

using namespace std;

int* load_primes(int n)
{
    int p, i, j;
    bool is_it_prime[n + 1];
    memset(is_it_prime, true, n + 1);

    for (p = 2; p * p <= n; p++) {
        if (is_it_prime[p] == true) {
            for (i = p * p; i <= n; i+= p) {
                is_it_prime[i] = false;
            }
        }
    }

    int* primes = new int[n];

    for (i = 2, j = 0; i <= n; i++) {
        if (is_it_prime[i] == true) {
            primes[j] = i;
            j++;
        }
    }

    return primes;
}

int josephus(int n, int k, int primes[])
{
    if (n == 1) {
        return 1;
    } else {
        return ((josephus(n - 1, k + 1, primes) + primes[k] - 1) % n + 1);
    }
}

int main()
{
    int n;
    int* primes = load_primes(32625);

    while (true) {
        cin >> n;

        if (n == 0) {
            break;
        }

        cout << (josephus(n, 0, primes)) << endl;
    }

    delete[] primes;
    return 0;
}

File: py/1032.py
from sys import setrecursionlimit

setrecursionlimit(10**5)


def loadPrimes(n):
    prime = [True for i in range(n + 1)]
    p = 2

    while (p * p <= n):
        if (prime[p] == True):

            for i in range(p ** 2, n + 1, p):

                prime[i] = False
        p += 1

    prime[0] = False
    prime[1] = False

    primes = [i for i in range(n + 1) if prime[i] == True]

    return primes


def josephus(n, k, primes):
    if n == 1:
        return 1

    else:
        return (josephus(n - 1, k + 1, primes) + primes[k] - 1) % n + 1


primes = loadPrimes(32625)

while True:
    n = int(input())

    if n == 0:
        break

    print(josephus(n, 0, primes))